File: transpiler/src/test/java/com/here/ivi/api/validator/SerializationValidatorPredicateTest.java
/*
 * Copyright (C) 2018 HERE Global B.V. and its affiliate(s). All rights reserved.
 *
 * This software, including documentation, is protected by copyright controlled by
 * HERE Global B.V. All rights are reserved. Copying, including reproducing, storing,
 * adapting or translating, any or all of this material requires the prior written
 * consent of HERE Global B.V. This material also contains confidential information,
 * which may not be disclosed to others without prior written consent of HERE Global B.V.
 *
 */

package com.here.ivi.api.validator;

import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import com.here.ivi.api.model.franca.FrancaDeploymentModel;
import org.franca.core.franca.*;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

@RunWith(JUnit4.class)
@SuppressWarnings("MethodName")
public class SerializationValidatorPredicateTest {

  @Mock private FModel fModel;
  @Mock private FTypeCollection francaTypeCollection;
  @Mock private FStructType francaStructType;
  @Mock private FField francaField;
  @Mock private FTypeRef francaTypeRef;

  @Mock private FrancaDeploymentModel deploymentModel;

  private final SerializationValidatorPredicate validatorPredicate =
      new SerializationValidatorPredicate();

  @Before
  public void setUp() {
    MockitoAnnotations.initMocks(this);

    when(fModel.getName()).thenReturn("");
    when(francaTypeCollection.getName()).thenReturn("");
    when(francaStructType.getName()).thenReturn("");
    when(francaField.getName()).thenReturn("");

    when(francaTypeCollection.eContainer()).thenReturn(fModel);
    when(francaStructType.eContainer()).thenReturn(francaTypeCollection);
    when(francaField.eContainer()).thenReturn(francaStructType);

    when(francaField.getType()).thenReturn(francaTypeRef);
    when(francaTypeRef.getPredefined()).thenReturn(FBasicTypeId.UNDEFINED);

    when(deploymentModel.isSerializable(francaStructType)).thenReturn(true);
  }

  @Test
  public void validateWithFieldInNonSerializableStruct() {
    when(deploymentModel.isSerializable(francaStructType)).thenReturn(false);

    assertNull(validatorPredicate.validate(deploymentModel, francaField));
  }

  @Test
  public void validateWithPrimitiveType() {
    when(francaTypeRef.getPredefined()).thenReturn(FBasicTypeId.BOOLEAN);

    assertNull(validatorPredicate.validate(deploymentModel, francaField));
  }

  @Test
  public void validateWithUndefined() {
    assertNotNull(validatorPredicate.validate(deploymentModel, francaField));
  }

  @Test
  public void validateWithTypeDefToUndefined() {
    FTypeRef undefinedTypeRef = mock(FTypeRef.class);
    when(undefinedTypeRef.getPredefined()).thenReturn(FBasicTypeId.UNDEFINED);
    FTypeDef francaTypeDef = mock(FTypeDef.class);
    when(francaTypeDef.getActualType()).thenReturn(undefinedTypeRef);
    when(francaTypeRef.getDerived()).thenReturn(francaTypeDef);

    assertNotNull(validatorPredicate.validate(deploymentModel, francaField));
  }

  @Test
  public void validateWithStructType() {
    when(francaTypeRef.getDerived()).thenReturn(mock(FStructType.class));

    assertNotNull(validatorPredicate.validate(deploymentModel, francaField));
  }

  @Test
  public void validateWithSerializableStructType() {
    FStructType serializableStruct = mock(FStructType.class);
    when(deploymentModel.isSerializable(serializableStruct)).thenReturn(true);
    when(francaTypeRef.getDerived()).thenReturn(serializableStruct);

    assertNull(validatorPredicate.validate(deploymentModel, francaField));
  }
}

File: franca-loader/src/test/java/com/here/genium/validator/SerializationValidatorPredicateTest.kt
/*
 * Copyright (C) 2016-2019 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.genium.validator

import com.here.genium.franca.FrancaDeploymentModel
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.mockk
import org.franca.core.franca.FBasicTypeId
import org.franca.core.franca.FField
import org.franca.core.franca.FModel
import org.franca.core.franca.FStructType
import org.franca.core.franca.FTypeCollection
import org.franca.core.franca.FTypeDef
import org.franca.core.franca.FTypeRef
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4

@RunWith(JUnit4::class)
class SerializationValidatorPredicateTest {
    @MockK
    private lateinit var fModel: FModel
    @MockK
    private lateinit var francaTypeCollection: FTypeCollection
    @MockK
    private lateinit var francaStructType: FStructType
    @MockK
    private lateinit var francaField: FField
    @MockK
    private lateinit var francaTypeRef: FTypeRef

    @MockK
    private lateinit var deploymentModel: FrancaDeploymentModel

    private val validatorPredicate = SerializationValidatorPredicate()

    @Before
    fun setUp() {
        MockKAnnotations.init(this, relaxed = true)

        every { fModel.name } returns ""
        every { francaTypeCollection.name } returns ""
        every { francaStructType.name } returns ""
        every { francaField.name } returns ""

        every { francaTypeCollection.eContainer() } returns fModel
        every { francaStructType.eContainer() } returns francaTypeCollection
        every { francaField.eContainer() } returns francaStructType

        every { francaField.type } returns francaTypeRef
        every { francaTypeRef.derived } returns null
        every { francaTypeRef.predefined } returns FBasicTypeId.UNDEFINED

        every { deploymentModel.isSerializable(francaStructType) } returns true
    }

    @Test
    fun validateWithFieldInNonSerializableStruct() {
        every { deploymentModel.isSerializable(francaStructType) } returns false

        assertNull(validatorPredicate.validate(deploymentModel, francaField))
    }

    @Test
    fun validateWithPrimitiveType() {
        every { francaTypeRef.predefined } returns FBasicTypeId.BOOLEAN

        assertNull(validatorPredicate.validate(deploymentModel, francaField))
    }

    @Test
    fun validateWithUndefined() =
        assertNotNull(validatorPredicate.validate(deploymentModel, francaField))

    @Test
    fun validateWithTypeDefToUndefined() {
        val undefinedTypeRef = mockk<FTypeRef>(relaxed = true)
        every { undefinedTypeRef.derived } returns null
        every { undefinedTypeRef.predefined } returns FBasicTypeId.UNDEFINED
        val francaTypeDef = mockk<FTypeDef>()
        every { francaTypeDef.actualType } returns undefinedTypeRef
        every { francaTypeRef.derived } returns francaTypeDef

        assertNotNull(validatorPredicate.validate(deploymentModel, francaField))
    }

    @Test
    fun validateWithStructType() {
        every { francaTypeRef.derived } returns mockk<FStructType>()

        assertNotNull(validatorPredicate.validate(deploymentModel, francaField))
    }

    @Test
    fun validateWithSerializableStructType() {
        val serializableStruct = mockk<FStructType>()
        every { deploymentModel.isSerializable(serializableStruct) } returns true
        every { francaTypeRef.derived } returns serializableStruct

        assertNull(validatorPredicate.validate(deploymentModel, francaField))
    }
}
